tests/test_auto_offer_on_3_device_rest.py

```diff
@@ -0,0 +1,146 @@
+"""
+Test file for the device client (Load, PV, and storage). Depends on d3a test setup file strategy_tests.external_devices
+"""
+import os
+import json
+import logging
+import sys
+import argparse
+from time import sleep
+from random import randint
+from d3a_api_client.rest_device import RestDeviceClient
+from d3a_api_client.utils import get_area_uuid_from_area_name_and_collaboration_id
+# parse arguments
+parser = argparse.ArgumentParser()
+parser.add_argument("--simulation_id", type=str, help="Simulation uuid")
+parser.add_argument("--load_names", nargs='+')
+parser.add_argument("--pv_names", nargs='+')
+parser.add_argument("--storage_names", nargs='+')
+args = parser.parse_args()
+
+connected_devices = []
+
+
+class AutoOfferBidOnMarket(RestDeviceClient):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.is_finished = False
+
+    def on_market_cycle(self, market_info):
+        """
+        Places a bid or an offer whenever a new market is created. The amount of energy
+        for the bid/offer depends on the available energy of the PV, or on the required
+        energy of the load.
+        :param market_info: Incoming message containing the newly-created market info
+        :return: None
+        """
+        if self.registered is False or self.is_finished is True:
+            return
+        logging.debug(f"New market information {market_info}")
+        logging.info("***********Market Info***************")
+        logging.info(market_info["device_info"])
+        logging.info("*************************************")
+        
+        if "available_energy_kWh" in market_info["device_info"] and market_info["device_info"]["available_energy_kWh"] > 0.0:
+            offered_energy = market_info["device_info"]["available_energy_kWh"]
+            offered_rate = 28*offered_energy
+            offer_response = self.offer_energy_rate(offered_energy, offered_rate)
+            expected_price = offered_rate*offered_energy
+            logging.info("*****************Offer Placed in Market by PV***************")
```

Comment:
In order to make this script applicable for multiple devices, I suggest to print the name of the device here and in the other places for better debugging. Reducing the number of `logger.info` calls will also make the debugging easier because there are less logging messages then.

Reply:
by device name, you mean which are passed as the argument ..right?

Reply:
I thought the device name is sent in the market_info. But it is not unfortunately.
You can distinguish the devices between each other with the `area_uuid`, which should be available in the market_info. So I wold be in favor of printing the area_uuid when printing the offer/bid placed message.